### PR TITLE
resinhup and dockerhub

### PR DESCRIPTION
**Pull request by agherzan**
Connected to #28

### CODE REVIEW

File: modules/util.py [hunk 1]
```diff
@@ -642,5 +649,19 @@ def testSafeDirCopyFile(self):
         dst = "./modules/util/safedircopy/dir3"
         self.assertFalse(safeDirCopy(src, dst))
 
+    def testSafeDirCopyIgnoreDir(self):
+        src = "./modules/util/safedircopy/dir1"
+        dst = "./modules/util/safedircopy/dir3"
+        self.assertTrue(safeDirCopy(src, dst, ignore=['ignore-dir']))
```
Comment (floion) [inline modules/util.py]: Why not use the safeCopy instead here?

File: modules/util.py [hunk 2]
```diff
@@ -642,5 +649,19 @@ def testSafeDirCopyFile(self):
...
+        self.assertFalse(os.path.isdir(os.path.join(dst, "ignore-dir")))
+        shutil.rmtree(dst) # cleanup
+
+    def testSafeDirCopyIgnoreFile(self):
+        src = "./modules/util/safedircopy/dir1"
+        dst = "./modules/util/safedircopy/dir3"
+        self.assertTrue(safeDirCopy(src, dst, ignore=['ignore-file']))
```
Comment (floion) [inline modules/util.py]: Why not use safeCopy instead here?

File: conf/resinhup.conf [hunk 3]
```diff
@@ -53,7 +53,9 @@ to_keep_files:
 [fetcher]
```
Comment (floion) [inline conf/resinhup.conf]: Can you rephrase the commit log please? :)

File: modules/util.py (hunk 2 above)
Comment (agherzan) [inline modules/util.py]: It would have worked with that too. I will change. Thanks.

File: modules/util.py (hunk 1 above)
Comment (agherzan) [inline modules/util.py]: It would have worked with that too. I will change. Thanks.

File: conf/resinhup.conf (hunk 3 above)
Comment (agherzan) [inline conf/resinhup.conf]: Will try. :)

File: fetcher/dockerhub.py [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+import logging
+from docker import Client
+from fetcher.tar import tarFetcher
+from modules.util import *
+
+log = logging.getLogger(__name__)
+
+class dockerhubFetcher(tarFetcher):
+
+    def __init__ (self, conffile, version, remote):
```
Comment (floion) [inline fetcher/dockerhub.py]: Why don't you call the parent's class __init__ and override only what you need?

File: fetcher/fetcher.py [hunk 5]
```diff
@@ -0,0 +1,24 @@
+#!/usr/bin/env python3
```
Comment (floion) [inline fetcher/fetcher.py]: Why not use this class to move the bulk from the tarFetcher class here?

File: fetcher/dockerhub.py (hunk 4 above)
Comment (agherzan) [inline fetcher/dockerhub.py]: I wanted to change this at the end but then I was to lazy to do it. I will have it. Given that this state was tested as it is, would it be fine to have a top commit for this?

File: fetcher/fetcher.py (hunk 5 above)
Comment (agherzan) [inline fetcher/fetcher.py]: Because the methods in the tar fetcher are tighted to tar archiving. We can see what we can split out but I propose to do it as a separate PR. Cause that is more like a refactoring commit.

File: fetcher/dockerhub.py (hunk 4 above)
Comment (floion) [inline fetcher/dockerhub.py]: Either way is fine

File: fetcher/fetcher.py (hunk 5 above)
Comment (floion) [inline fetcher/fetcher.py]: As you consider best

File: fetcher/dockerhub.py (hunk 4 above)
Comment (agherzan) [inline fetcher/dockerhub.py]: Doing.

File: conf/resinhup.conf (hunk 3 above)
Comment (floion) [inline conf/resinhup.conf]: "The resinhup update package have changes the format" - this still throws me off

File: scripts/docker-build-and-push.sh [hunk 6]
```diff
@@ -77,7 +77,7 @@ for dockerfile in $DOCKERFILES; do
     fi
     printf "${GREEN}Running build for $device using $dockerfile ...${NC}\n"
     docker build -t resinhup-$device:$TAG -f ../$dockerfile $SCRIPTPATH/..
-    printf "${GREEN}Tag an push for $device ...${NC}\n"
+    printf "${GREEN}Tag and push for $device ...${NC}\n"
```
Comment (floion) [inline scripts/docker-build-and-push.sh]: Typo in the commit log also: "sypo" there

Comment (agherzan) [inline scripts/docker-build-and-push.sh]: Fixed

File: conf/resinhup.conf (hunk 3 above)
Comment (agherzan) [inline conf/resinhup.conf]: :) Fixed.

**Comment by floion** [comment]
LGTM

File: fetcher/dockerhub.py [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+        super().__init__()
+        self.remotefile = os.path.join(self.remote + ":resinhup-" + machine + "-" + version)
```
Comment (petrosagg) [inline fetcher/dockerhub.py]: As commented here https://github.com/resin-os/resin-yocto-scripts/pull/33#r79108514 , let's change the tag to be without `resinhup-`

File: fetcher/dockerhub.py [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+            cli.remove_container(container_name, force=True)
+        except:
+            pass
+
+        # Create resinhup container
+        try:
+            # Docker mounts a set of files at container creation
+            # https://docs.docker.com/engine/userguide/networking/default_network/configure-dns/
+            # We still need all these files as they are important in the update process
+            cmd='bash -c "umount /etc/resolv.conf ; umount /etc/hostname ; umount /etc/hosts"'
```
Comment (petrosagg) [inline fetcher/dockerhub.py]: This has no effect on the export

File: fetcher/dockerhub.py [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+            container = cli.create_container(image=self.remotefile, command=cmd, name=container_name)
+        except:
+            log.error("Can't create temporary update container.")
+            return False
+
+        # Run container
+        try:
+            response = cli.start(container=container.get('Id'), privileged=True)
```
Comment (petrosagg) [inline fetcher/dockerhub.py]: No need to run any commands or start the container. You can export right after creating

File: fetcher/dockerhub.py [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+
+        # Export the temporary resinhup container as tar archive
+        try:
+            strm = cli.export(container=container.get('Id'))
+        except:
+            log.error("Can't export tar archive update file.")
+            return False
+        with open(self.workspacefile, 'wb') as fd:
+            for chunk in strm.stream(1000000):
+                fd.write(chunk)
```
Comment (petrosagg) [inline fetcher/dockerhub.py]: Adhering to how the original tar fetcher works means that now we need double space in the data partition. This would be much better if it passed the stream to `tarfile` directly

File: fetcher/tar.py [hunk 11]
```diff
@@ -99,18 +101,26 @@ def unpack(self, downloadFirst=False):
         update = tarfile.open(name=self.workspacefile, mode='r:*')
```
Comment (petrosagg) [inline fetcher/tar.py]: Instead of passing a file here you can replace with a file object. Here is how you would open the stream returned by dockerpy:

```python
import docker
import tarfile

cli = docker.Client()
ctnr = cli.create_container('busybox', 'true')
strm = cli.export(ctnr)
tar = tarfile.open(mode='r|*', fileobj=strm)
tar.list()
```

The important bit is the `r|*` (instead of `r:*`) mode which tells tarfile to not perform any `seek()` operations and therefore can operate on a stream

File: fetcher/dockerhub.py [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+        # Run container
+        try:
+            response = cli.start(container=container.get('Id'), privileged=True)
+        except:
+            log.error("Can't start temporary update container.")
+            return False
+
+        # Export the temporary resinhup container as tar archive
+        try:
+            strm = cli.export(container=container.get('Id'))
```
Comment (petrosagg) [inline fetcher/dockerhub.py]: You can pass the `container` variable directly

File: fetcher/tar.py [hunk 13]
```diff
@@ -99,18 +101,26 @@ def unpack(self, downloadFirst=False):
...
         update.extractall(self.workspaceunpack)
```
Comment (petrosagg) [inline fetcher/tar.py]: Why is `workspaceunpack` not the unused root partition? This means that the current PR needs 3*(root_size) of free space available in the data partition.

File: fetcher/tar.py [hunk 14]
```diff
@@ -99,18 +101,26 @@ def unpack(self, downloadFirst=False):
...
-        rootfsfiles = os.listdir(self.workspace + "/update/resin-root")
-        self.rootfsarchive = self.workspace + "/update/resin-root/" + os.listdir(self.workspace + "/update/resin-root")[0]
-
         log.debug("Unpacked " + self.workspacefile + " in " + self.workspaceunpack)
         return True
 
+    def unpackQuirks(self, location):
+        '''Copies the content of <quirks> directory from an update to <location>.'''
+        quirks_path = os.path.join(self.workspaceunpack, 'quirks')
+        log.info("Unpacking rootfs quirks from " + quirks_path + " to " + self.workspaceunpack + ".")
```
Comment (petrosagg) [inline fetcher/tar.py]: According to the code, unpacking happens from `quirks_path` to `location`, not `self.workspaceunpack`

File: fetcher/tar.py [hunk 15]
```diff
@@ -99,18 +101,26 @@ def unpack(self, downloadFirst=False):
...
+            return False
+        log.info("Unpack rootfs quirks done.")
+        return True
+
     def unpackRootfs(self, location):
         log.info("Unpack rootfs started... this can take a couple of seconds or even minutes...")
-        update = tarfile.open(self.rootfsarchive)
-        update.extractall(location)
-        log.debug("Unpacked rootfs " + self.rootfsarchive + " in " + location)
+        safeCopy(self.workspaceunpack, location, sync=False, ignore=['resin-boot', 'quirks'])
```
Comment (petrosagg) [inline fetcher/tar.py]: Why don't need the ignore mechanism at all if the operations happen in the right order. Can we only unpack __once__ into the inactive root partition and then do the two copy operations for `resin-boot` and `quirks` from there?

File: fetcher/tar.py (hunk 11 above)
Comment (floion) [inline fetcher/tar.py]: Nice. This would also mean some small changes in the tarFetcher class to define a boolean that would signal if we get the stream from the dockerhubFetcher or prepare it ourselves here in the unpack() method.

File: fetcher/dockerhub.py (hunk 7 above)
Comment (agherzan) [inline fetcher/dockerhub.py]: Deal.

File: fetcher/dockerhub.py (hunk 8 above)
Comment (agherzan) [inline fetcher/dockerhub.py]: This was an old implementation for the fix of hostname and co shadowing. I forgot to remove it. Good catch.

File: fetcher/dockerhub.py (hunk 9 above)
Comment (agherzan) [inline fetcher/dockerhub.py]: This was an old implementation for the fix of hostname and co shadowing. I forgot to remove it. Good catch.

File: fetcher/dockerhub.py (hunk 12 above)
Comment (agherzan) [inline fetcher/dockerhub.py]: The container variable holds a dict. While the export expects a string as the container argument `container (str): The container to export`. Did you mean `container_name`?

File: fetcher/dockerhub.py (hunk 10 above)
Comment (agherzan) [inline fetcher/dockerhub.py]: That makes a lot of sense.

File: fetcher/tar.py (hunk 11 above)
Comment (agherzan) [inline fetcher/tar.py]: @petrosagg Definitely a very good point.
@floion No. Actually all we need is to replace workspacefile by IO stream.

File: fetcher/tar.py (hunk 14 above)
Comment (agherzan) [inline fetcher/tar.py]: Type. Fixing.

File: fetcher/tar.py (hunk 15 above)
Comment (agherzan) [inline fetcher/tar.py]: See comment above about rootfs size.

File: fetcher/dockerhub.py (hunk 12 above)
Comment (floion) [inline fetcher/dockerhub.py]: As Petros pointed out in a comment in fetcher/tar.py, the following piece of code does the export using the variable returned by create_container:

import docker
import tarfile

cli = docker.Client()
ctnr = cli.create_container('busybox', 'true')
strm = cli.export(ctnr)
tar = tarfile.open(mode='r|*', fileobj=strm)
tar.list()

File: fetcher/tar.py (hunk 11 above)
Comment (floion) [inline fetcher/tar.py]: Well, if you also want to keep the tarFetcher, you will need to adapt it to work with streams I think to better accommodate both cases. So it can work with such a boolean, plus creating a stream out of the tar archive the tar fetcher downloads

File: fetcher/tar.py (hunk 13 above)
Comment (floion) [inline fetcher/tar.py]: Why not? Isn't the unused rootfs formatted by resinhup at the beginning of the process?

Comment (agherzan) [inline fetcher/tar.py]: @floion the resinhup update packages (the archive) containes rootfs + boot + some
files. Extracting this in rootfs will most probably go over the rootfs size.

File: fetcher/tar.py (hunk 11 above)
Comment (agherzan) [inline fetcher/tar.py]: Indeed changes are needed but I don't need a flag for this. I can simply use a stream. Implemented in the last push.

File: fetcher/dockerhub.py (hunk 12 above)
Comment (agherzan) [inline fetcher/dockerhub.py]: It is just an undocumented way of using it. It is handled by code https://github.com/docker/docker-py/blob/c5a81002c6af649ed5deef44b154b05627559e1b/docker/utils/decorators.py#L7 but the documentation clearly states that it needs a str as an argument.

File: fetcher/tar.py (hunk 13 above)
Comment (agherzan) [inline fetcher/tar.py]: There is a problem too in extracting the boot files directly from the tar to the boot partition: handling atomicity which is currently handled as a copy method in resinhup.

File: modules/updater.py [hunk 16]
```diff
@@ -141,7 +146,7 @@ def rootfsOverlay(self):
             if not os.path.exists(src_full_path):
                 log.warn(src_full_path + " was not found in your current mounted rootfs. Can't overlay.")
                 continue
-            if not safeCopy(src_full_path, self.tempRootMountpoint + dst):
+            if not safeCopy(src_full_path, self.tempRootMountpoint + dst, sync=False):
```
Comment (petrosagg) [inline modules/updater.py]: I don't think this is a necessary change especially since it introduces a subtle way of breaking the updater which is the power cutting after having switched the bootloader but before these writes have been persisted. Since performance isn't an issue I suggest keeping all the operations safe

File: fetcher/tar.py [hunk 17]
```diff
@@ -59,58 +59,58 @@ def download(self):
             log.error("HTTP status code: " + str(r.status_code))
             return False
 
-        with open(self.workspacefile, 'wb') as fd:
-            for chunk in r.iter_content(1000000):
-                fd.write(chunk)
+        self.updatefilestream = StringIO(r.content)
```
Comment (petrosagg) [inline fetcher/tar.py]: This is a really bad idea. It will load the whole tar archive into memory.

File: fetcher/tar.py (hunk 15 above)
Comment (petrosagg) [inline fetcher/tar.py]: Ok. Understood. This is something to fix in the next releases.

File: modules/updater.py (hunk 16 above)
Comment (agherzan) [inline modules/updater.py]: Got it. Sure.
